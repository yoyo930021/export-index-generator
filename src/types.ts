import type { SourceFile, Program } from 'typescript'

export interface FileProgram {
  ast: SourceFile
  program: Program
}

export interface ExportModule {
  name: string
  isTypeOnly: boolean
  default: boolean
}
