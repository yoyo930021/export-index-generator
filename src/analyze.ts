import { FileProgram, ExportModule } from './types'
import type ts from 'typescript'
import { basename, extname } from 'path'

export const guessIsType = (tsModule: typeof ts, typeChecker: ts.TypeChecker, symbol: ts.Symbol): boolean => {
  const declaration = symbol.valueDeclaration ?? symbol.declarations?.[0]
  if (!declaration) {
    throw new Error(`No declaration for ${JSON.stringify(symbol)}`)
  }

  // export interface ex6 {}
  if (tsModule.isInterfaceDeclaration(declaration)) {
    return true
  }

  // export type ex7 = '1'
  if (tsModule.isTypeAliasDeclaration(declaration)) {
    return true
  }

  // export { ex8, ex9 }
  if (tsModule.isExportSpecifier(declaration)) {
    const originalSymbol = typeChecker.getAliasedSymbol(symbol)
    if (!(originalSymbol.valueDeclaration ?? originalSymbol.declarations?.[0])) return false
    return guessIsType(tsModule, typeChecker, originalSymbol)
  }

  // export default ex10
  if (tsModule.isExportAssignment(declaration)) {
    const originalSymbol = typeChecker.getAliasedSymbol(symbol)
    if (!(originalSymbol.valueDeclaration ?? originalSymbol.declarations?.[0])) return false
    return guessIsType(tsModule, typeChecker, originalSymbol)
  }

  return false
}

export const getDefaultName = (tsModule: typeof ts, symbol: ts.Symbol, fileName: string): string => {
  const declaration = symbol.valueDeclaration ?? symbol.declarations?.[0]
  if (!declaration) {
    throw new Error(`No declaration for ${JSON.stringify(symbol)}`)
  }

  if (tsModule.isClassDeclaration(declaration)) {
    return declaration.name?.escapedText as string || fileName
  }

  if (tsModule.isExportAssignment(declaration)) {
    return declaration.expression?.getText() || fileName
  }

  if (tsModule.isInterfaceDeclaration(declaration)) {
    return declaration.name.escapedText as string || fileName
  }

  return fileName
}

export const analyzeExports = (tsModule: typeof ts, fileProgram: FileProgram, allExportedNames: Set<string>): ExportModule[] => {
  const { ast, program } = fileProgram
  const typeChecker = program.getTypeChecker()

  const symbol = typeChecker.getSymbolAtLocation(ast)
  if (!symbol || !symbol.exports) return []

  return Array.from(symbol.exports.values() as unknown as Iterable<ts.Symbol>)
    .map((item) => {
      if (item.name === 'default') {
        return {
          name: getDefaultName(tsModule, item, basename(ast.fileName, extname(ast.fileName))),
          isTypeOnly: guessIsType(tsModule, typeChecker, item),
          default: true
        }
      }
      if (allExportedNames.has(item.name)) {
        throw new Error(`Duplicate export name found: ${item.name} in file ${ast.fileName}`)
      }
      return { name: item.name, isTypeOnly: guessIsType(tsModule, typeChecker, item), default: false }
    })
}

export const getExternalExports = (path: string): ExportModule[] => {
  return [{
    name: basename(path).split('.')[0],
    isTypeOnly: false,
    default: true
  }]
}
