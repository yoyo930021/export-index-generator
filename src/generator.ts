import { ExportModule } from './types'
import type ts from 'typescript'
import { supportExtensions } from './constant'
import { analyzeExports, getExternalExports } from './analyze'
import { getSingleFileProgram } from './parser'
import { basename, extname } from 'path'

const removeNoNeedExtension = (path: string): string => {
  for (const ext of supportExtensions) {
    const result = path.replace(new RegExp(`\\.${ext}$`, 'g'), '')
    if (result !== path) return result
  }

  return path
}

const generateExports = (tsModule: typeof ts, path: string, exports: ExportModule[]): ts.ExportDeclaration[] => {
  const result: ts.ExportDeclaration[] = []
  const isTypeOnlyExports = exports.filter((item) => item.isTypeOnly)
  const otherExports = exports.filter((item) => !item.isTypeOnly)

  const createExportSpecifier = (item: ExportModule) => {
    const [major, minor] = tsModule.version.split('.')
    const createIdentifier = (text: string): ts.Identifier => {
      if (Number(major) === 5) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-expect-error
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
        return tsModule.factory.createIdentifier(text)
      }
      return tsModule.createIdentifier(text)
    }
    const propertyName = item.default ? createIdentifier('default') : undefined

    if ((Number(major) === 5)) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-return
      return tsModule.factory.createExportSpecifier(undefined, propertyName, tsModule.factory.createIdentifier(item.name))
    }
    if ((Number(major) === 4 && Number(minor) >= 5)) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      return tsModule.createExportSpecifier(undefined, propertyName, tsModule.createIdentifier(item.name))
    }
    return tsModule.createExportSpecifier(propertyName, tsModule.createIdentifier(item.name))
  }

  const createExportDeclaration = (input: ExportModule[], isOnlyType: boolean) => {
    const [major] = tsModule.version.split('.')
    if ((Number(major) === 5)) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
      return tsModule.factory.createExportDeclaration(
        undefined,
        isOnlyType,
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-expect-error
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
        tsModule.factory.createNamedExports(
          input.map(createExportSpecifier)
        ),
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-expect-error
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
        tsModule.factory.createStringLiteral(`./${removeNoNeedExtension(path)}`)
      )
    }
    return tsModule.createExportDeclaration(
      undefined,
      undefined,
      tsModule.createNamedExports(
        input.map(createExportSpecifier)
      ),
      tsModule.createStringLiteral(`./${removeNoNeedExtension(path)}`),
      isOnlyType
    )
  }

  if (isTypeOnlyExports.length > 0) {
    result.push(createExportDeclaration(isTypeOnlyExports, true))
  }

  if (otherExports.length > 0) {
    result.push(createExportDeclaration(otherExports, false))
  }

  return result
}

export interface File {
  path: string
  content: string
}

export const generate = (tsModule: typeof ts, files: File[]): { content: string, scriptKind: ts.ScriptKind } => {
  const scriptKind = (() => {
    if (files.some((file) => extname(file.path) === '.tsx')) return tsModule.ScriptKind.TSX
    if (files.some((file) => extname(file.path) === '.ts')) return tsModule.ScriptKind.TS
    if (files.some((file) => extname(file.path) === '.jsx')) return tsModule.ScriptKind.JSX
    return tsModule.ScriptKind.JS
  })()

  const sourceFile = tsModule.createSourceFile(
    'index.ts',
    '',
    tsModule.ScriptTarget.ESNext,
    true,
    scriptKind
  )

  const statements = files.reduce(
    (result, file) => {
      // External
      if (!supportExtensions.some((ext) => file.path.endsWith(ext))) return result.concat(generateExports(tsModule, file.path, getExternalExports(file.path)))

      // Script
      const exports = analyzeExports(tsModule, getSingleFileProgram(tsModule, basename(file.path), file.content))
      return result.concat(generateExports(tsModule, file.path, exports))
    },
    [] as ts.ExportDeclaration[]
  )

  const printer = tsModule.createPrinter()
  const updateSourceFileNode = (sourceFile: ts.SourceFile, statements: ts.ExportDeclaration[]) => {
    const [major] = tsModule.version.split('.')
    if ((Number(major) === 5)) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
      return tsModule.factory.updateSourceFile(sourceFile, statements)
    }
    return tsModule.updateSourceFileNode(sourceFile, statements)
  }

  return {
    content: printer.printFile(updateSourceFileNode(sourceFile, statements)).replace(/;/g, '').replace(/"/g, '\''),
    scriptKind
  }
}
