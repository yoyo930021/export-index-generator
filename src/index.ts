import { program } from 'commander'
import fg from 'fast-glob'
import { supportExtensions } from './constant'
import ts from 'typescript'
import { generate } from './generator'
import { resolve } from 'path'
import { readFileSync, writeFileSync } from 'fs'

const getCurrentCommand = (external: boolean, cwd: string) => `generate-index${external ? ' -e' : ''} ${cwd}`

const getFileName = (tsModule: typeof ts, scriptKind: ts.ScriptKind) => {
  if (scriptKind === tsModule.ScriptKind.TSX) return 'tsx'
  if (scriptKind === tsModule.ScriptKind.TS) return 'ts'
  if (scriptKind === tsModule.ScriptKind.JSX) return 'jsx'
  return 'js'
}

const addAutoComment = (command: string, content: string) => `/* auto generator by \`${command}\` */
${content}/* auto generator end */
`

const readFile = (path: string) => {
  try {
    return readFileSync(path, { encoding: 'utf8' })
  } catch {
    return null
  }
}

const updateOrCreateFile = (command: string, path: string, content: string) => {
  const oldContent = readFile(path) ?? ''
  const startRe = /\/\* auto generator by.* \*\/\n/
  const endRe = /\/\* auto generator end \*\/\n/
  const start = startRe.exec(oldContent)
  const end = endRe.exec(oldContent)

  const output =
    (start && end)
      ? oldContent.slice(0, start.index + start[0].length) + content + oldContent.slice(end.index)
      : addAutoComment(command, content)

  console.log((start && end) ? 'Update file.' : 'Create new file.')
  writeFileSync(path, output)
}

const successMessage = 'âœ¨ Success'
program
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  .version((require('../package.json') as { version: string }).version)
  .arguments('<cwd>')
  .description('generate typescript or javascript index file')
  .option('-e, --external', 'generate no script type file with fileName', false)
  .option('-d, --detect-duplicate', 'detect duplicate exports', false)
  .action((cwd: string, cmd: { external: boolean, detectDuplicate: boolean }) => {
    console.time(successMessage)
    const cwdPath = resolve(process.cwd(), cwd)
    const tsModule = (() => {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      try { return require(require.resolve('typescript', { paths: [cwdPath] })) as typeof ts } catch { return ts }
    })()

    const extension = cmd.external ? '*' : `{${supportExtensions.join(',')}}`
    const files =
      fg
        .sync(`**/*.${extension}`, { cwd: cwdPath }).filter((file) => !/^index\.[jt]{1}sx{0,1}$/.test(file))
        .map((path) => ({ path, content: readFileSync(resolve(cwd, path), { encoding: 'utf8' }) }))

    const result = generate(tsModule, files, cmd.detectDuplicate)
    updateOrCreateFile(
      getCurrentCommand(!!cmd.external, cwd),
      resolve(cwd, `index.${getFileName(tsModule, result.scriptKind)}`),
      result.content
    )
    console.timeEnd(successMessage)
  })

program.parse(process.argv)
