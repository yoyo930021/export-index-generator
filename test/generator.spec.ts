import { createRequire } from 'module'
import { describe, expect, it } from 'vitest'
import { generate } from '../src/generator'
import { supportExtensions } from '../src/constant'
import fg from 'fast-glob'
import { resolve } from 'path'
import { readFileSync } from 'fs'

const folders = ['fixture/ex1']
const requireModule = createRequire(__filename)

type TsModule = typeof import('typescript')

const loadTypeScript = (specifier: string): TsModule =>
  requireModule(specifier) as TsModule

const typeScriptVariants: Array<{ label: string, ts: TsModule }> = [
  { label: 'TypeScript 3.9', ts: loadTypeScript('typescript') },
  { label: 'TypeScript 4.9', ts: loadTypeScript('typescript-4-9') }
]

for (const { label, ts } of typeScriptVariants) {
  describe(label, () => {
    it('generate []', () => {
      expect(generate(ts, [])).toMatchSnapshot()
    })

    for (const folder of folders) {
      it(`generate ${folder}`, () => {
        const cwd = resolve(__dirname, folder)
        const files =
          fg
            .sync(`**/*.{${supportExtensions.join(',')}}`, { cwd })
            .filter((file) => !/^index\.[jt]{1}sx{0,1}$/.test(file))
            .map((path) => ({
              path,
              content: readFileSync(resolve(cwd, path), { encoding: 'utf8' })
            }))

        expect(generate(ts, files)).toMatchSnapshot()
      })
    }
  })
}
