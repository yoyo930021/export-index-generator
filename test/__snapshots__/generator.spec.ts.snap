// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TypeScript 3.9 > generate [] 1`] = `
{
  "content": "",
  "scriptKind": 1,
}
`;

exports[`TypeScript 3.9 > generate fixture/ex1 1`] = `
{
  "content": "export { Test5, Test6, default as Test7 } from './class'
export { Test16, Test17 } from './enum'
export type { Test13, default as Test15 } from './export'
export { Test14 } from './export'
export { test8, test9, test10 } from './function'
export type { Test2, Test3, default as Test4 } from './interface'
export type { Test11, Test12 } from './type'
",
  "scriptKind": 3,
}
`;

exports[`TypeScript 4.9 > generate [] 1`] = `
{
  "content": "",
  "scriptKind": 1,
}
`;

exports[`TypeScript 4.9 > generate fixture/ex1 1`] = `
{
  "content": "export { Test5, Test6, default as Test7 } from './class'
export { Test16, Test17 } from './enum'
export type { Test13, default as Test15 } from './export'
export { Test14 } from './export'
export { test8, test9, test10 } from './function'
export type { Test2, Test3, default as Test4 } from './interface'
export type { Test11, Test12 } from './type'
",
  "scriptKind": 3,
}
`;
